content/3_Silver/Func_Graphs.mdx

```diff
@@ -444,51 +445,177 @@ Try simulating the algorithm on the following directed graph in CSAcademy's
 
 ### Solution 2
 
-`go(x)` generates answers for all vertices in the connected component containing
+`floyd(x)` generates answers for all vertices in the connected component containing
 `x`. Requires reverse adjacency lists (`radj`).
 
 <LanguageSection>
 
 <CPPSection>
 
 ```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
 int N;
-vi P,ans;
-V<vi> radj;
+vector<int> adj, ans;
+vector<vector<int>> radj;
 
-void fill_radj(int x) {
-	trav(t,radj[x]) if (ans[t] == -1) {
-		ans[t] = ans[x];
-		fill_radj(t);
+void fill_radj(int x)
+{
+	for (auto &child : radj[x])
+	{
+		// as all nodes in the cycle are processed in function floyd,
+		// the recursive call will only start at the nodes which
+		// combine the cycle with the acyclic part of the connected component
+		// where one of its outgoing arrows points to the node that is not processed yet
+		if (ans[child] == -1)
+		{
+			ans[child] = ans[x];
+			fill_radj(child);
+		}
 	}
 }
 
-void go(int x) {
+void floyd(int x)
+{
 	int y = x;
-	do { // find a cycle using Floyd's algo
-		x = P[x];
-		y = P[P[y]];
+	do
+	{ // find a cycle using Floyd's algorithm
+		x = adj[x];
+		y = adj[adj[y]];
 	} while (y != x);
-	do { // set ans[x] = x for all x along cycle
+	do
+	{ // set ans[x] = x for all x along cycle
 		ans[x] = x;
-		x = P[x];
+		x = adj[x];
 	} while (y != x);
-	do { // set ans'es for all x not along cycle
+	do
+	{ // set ans'es for all x not along cycle
 		fill_radj(x);
-		x = P[x];
+		x = adj[x];
 	} while (y != x);
 }
 
-int main() {
-	setIO(); rv(N,P); trav(t,P) --t;
-	radj.rsz(N); F0R(i,N) radj[P[i]].pb(i);
-	ans = vi(N,-1); F0R(i,N) if (ans[i] == -1) go(i);
-	trav(t,ans) pr(t+1,' ');
+int main()
+{
+	cin.tie(0)->sync_with_stdio(false);
+	cin >> N;
+
+	adj.assign(N, -1);
+	for (auto &e : adj)
+	{
+		cin >> e;
+		e--;
+	}
+
+	// we need vector of vectors since at nodes combining the cycle with other
+	// part of the connected component,
+	// there would be more than one outgoing arrows in the reversed adjacency list
+	radj.assign(N, {});
+	for (int i = 0; i < N; i++)
+		radj[adj[i]].push_back(i);
+
+	ans.assign(N, -1);
+	// we run the Floyd's algorithm for each connected component
+	for (int i = 0; i < N; i++)
+		if (ans[i] == -1)
+			floyd(i);
+
+	for (auto &a : ans)
+		cout << a + 1 << " ";
 }
 ```
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class Badge {
+	static Integer[] adj, ans;
+	/*
+	 * we need list of lists for the reversed adjacency list since at nodes
```

Comment:
```suggestion
	 * We need list of lists for the reversed adjacency list since at nodes
```
the `list of lists` part could probably be worded a little better (same with vector of vectors)

Reply:
probably something along the lines of "for each node store a vector" would seem more clear.